Lesson20/selcan/oop.js

```diff
@@ -0,0 +1,106 @@
+class Shape {
+    constructor(name,color){
+        this._name = name;
+        this._color =color;
+    }
+    description(){
+        return `This is a ${this._color} square.`
+    }
+}
+class Square extends Shape{
+    constructor(name,color,sideLength){
+        super(name,color);
+        this._sideLength = sideLength;
+        this._area = sideLength * sideLength;
+    }
+    description(){
+        return `This is a ${this._color} square with a side length of ${this._sideLength} and an area of ${this._area}.`
+    }
+}
+const square1 = new Square("kare","blue",20)
+const Shape1 = new Shape("kare","pink");
+console.log("1:",Shape1.description());
+console.log("2:",square1);
+console.log("3:",square1._area);
+console.log("4:",square1.description());
+console.log("5:",Shape1.description());
+
+
+
+
+
+// 3. Create a Vehicle class with make, model, and year properties, and a start() method that logs a message to the console indicating that the vehicle has started. Then create a Car class that extends the Vehicle class and adds a numDoors property. Override the start() method in the Car class to log a different message to the console indicating that the car has started.
+
+class Vehicle {
+    constructor(make,model,year){
+        this._make = make;
+        this._model = model;
+        this._year = year;
+    }
+    start(){
+        console.log(`${this._make} mali,${this._model} model Vehicle has started.`)
+    }
+}
+class Car extends Vehicle{
+    constructor(make,model,year){
+        super(make,model,year)
+    }
+    start(){
+        console.log(`${this._make} mali,${this._model} model car has started.`)
+    }
+}
+const Vehicle1 = new Vehicle("Çin","Ford",2016);
+const Car1 = new Car("Alman","Mercedes",2020)
+
+
+Vehicle1.start();
+Car1.start();
+
+
+
+
+// 4. Create a BankAccount class with balance and interestRate properties, and deposit() and withdraw() methods. Add a transactionHistory property that is an array of objects representing each transaction made on the account, with properties for type (either "deposit" or "withdrawal"), amount, and date ({type: "withdrawal", amount: amount, date: new Date().toLocaleDateString()}). Implement deposit() and withdraw() methods that add new transactions to the transactionHistory array. Implement a get transactionHistory() getter method that returns a copy of the transaction history array, and a get currentBalance() getter method that calculates and returns the current balance of the account based on the transaction history.
+
+class BankAccount{
+    constructor(balance,interestRate){
+        this._balance = balance;
+        this._interestRate = interestRate;
+        this._transactionHistory = [];
+
+    }
+    deposit(amount){
```

Comment:
Deposit method should look like this:
```

deposit(amount) {
        this._balance += amount;
        this._transactionHistory.push({ type: "deposit", amount: amount, date: new Date().toLocaleDateString() });
    }
```

Try adding withdraw method yourself but it will be the opposite of what deposit method do (reduce the balance, add new transaction to the transactionHistory array). Also, it would be good to check that there's enough money in the account (balance value) before doing the withdraw operation:
```
withdraw(amount) {
        if (amount > this._balance) {
            console.log("Insufficient funds, please try smaller amount.");
            return;
        }
    }
```